### PR TITLE
Fix #10068 - Progress Callback should report 100 when simulation completely succesfully

### LINKED ISSUE #10068: [pyenergyplus] notifications for warmup and simulation ends
Issue overview
--------------

In order to integrate EnergyPlus Python API in other programs, it's sometimes necessary to asynchronously consume variables and send actuators values. 
One example of such use case can be found here: https://github.com/airboxlab/rllib-energyplus/blob/main/run.py, which is essentially running EnergyPlus simulations in OpenAI Gym environment as part of a reinforcement learning experiment, backed by a distributed training framework (Ray). To integrate with the python API paradigm mainly based on callbacks, queues are used as exchange mechanisms. 

Some control flow mechanisms that are used and associated limitations:
-  As there is no callback signalling end of simulation (`callback_progress` may or may not send the 100% value), a timeout needs to be used on queues to overcome the problem of waiting forever when simulation ends.
- when run begins, the warmup period has to be executed before the first variable or meter value is received. This means a sufficiently long timeout has to be configured. The longer the warmup (or the timestep execution) the longer the timeout.
- as a consequence, you may wait quite long after the last call to callbacks has been issued, leading to performance and throughput issues in long-lived experiments involving many simulations.

A potential enhancement would be to introduce additional callbacks to notify about warmup and simulation phases status, most importantly when they are finished. With that, it's possible to introduce additional mechanisms to wait on warmup and clean stop when simulation is finished, and to reduce timeout on variable values reception to a minimum corresponding to the estimated duration of a timestep. This still requires configuration but it's more predictible and deterministic than warmup period duration.

### Details

Some additional details for this issue (if relevant):
 - Platform (Operating system, version): Ubuntu 20.04
 - Version of EnergyPlus: v22.2.0

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #10068

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/UtilityRoutines.cc [hunk 1]
```diff
@@ -776,6 +776,13 @@ int EndEnergyPlus(EnergyPlusData &state)
     // indicating that E+ finished its simulation
     if ((state.dataExternalInterface->NumExternalInterfaces > 0) && state.dataExternalInterface->haveExternalInterfaceBCVTB) CloseSocket(state, 1);
 
+    if (state.dataGlobal->fProgressPtr) {
+        state.dataGlobal->fProgressPtr(100);
+    }
+    if (state.dataGlobal->progressCallback) {
+        state.dataGlobal->progressCallback(100);
+    }
```
Comment (jmarrec) [inline src/EnergyPlus/UtilityRoutines.cc:784]: Register 100 in EndEnergyPlus when all is well.

File: tst/EnergyPlus/api/TestRuntime.c [hunk 2]
```diff
@@ -186,6 +188,9 @@ int main(int argc, const char *argv[])
         printf("There were %d warnings!\n", numWarnings);
         numWarnings = 0;
     }
+    if (progressValue != 100) {
+        return 1;
+    }
```
Comment (jmarrec) [inline tst/EnergyPlus/api/TestRuntime.c:193]: RuntimeC test, fails before fix

File: tst/EnergyPlus/api/TestRuntime.py [hunk 3]
```diff
@@ -91,6 +94,7 @@ def error_handler(severity: int, message: bytes) -> None:
 if v != 0:
     print("EnergyPlus Failed!")
     sys.exit(1)
+assert(progressValue == 100)
```
Comment (jmarrec) [inline tst/EnergyPlus/api/TestRuntime.py:97]: RuntimePython test, fails before fix

File: src/EnergyPlus/UtilityRoutines.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UtilityRoutines.cc:784]: Is this executed when E+ encounters a fatal and is aborting?

File: tst/EnergyPlus/api/TestRuntime.c (hunk 2 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/api/TestRuntime.c:193]: Good, ensures we get 100 when we leave.

File: tst/EnergyPlus/api/TestRuntime.py (hunk 3 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/api/TestRuntime.py:97]: :+1:

**Review by Myoldmopar — COMMENTED** [review]
This is all reasonable, my only question being that I can't remember if E+ goes through EndEnergyPlus if the simulation is aborting.  I think we should still send 100 either way, and I just can't remember if this covers that or not.

**Comment by Myoldmopar** [comment]
CI is all happy here, and this is an appropriate fix.  If we need additional work on the API callbacks, they can come over time, but I'm merging this.  Thanks for the quick fixup @jmarrec

**Comment by jmarrec** [comment]
> This is all reasonable, my only question being that I can't remember if E+ goes through EndEnergyPlus if the simulation is aborting. I think we should still send 100 either way, and I just can't remember if this covers that or not.

No it doesn't go through EndEnergyPlus. And I put it there on purpose. I  don't want 100% when it failed, I want whatever value at which it stopped.

**Comment by Myoldmopar** [comment]
Yep, that makes sense.  Thanks @jmarrec

**Pull request merged**